### PR TITLE
Fixed dependency resolution using Cargo.lock

### PR DESCRIPTION
**Pull request by leonardohn**
Parsing dependencies from Cargo.lock should overcome the version resolution limitations from `cargo-metadata` and also allow to use the same dep versions after a successful `cargo build`. It has been tested on both `cargo-ebuild` and `rust-analyzer`.

### CODE REVIEW

**Comment by cardoe** [comment]
What if there's no lock file?

**Comment by leonardohn** [comment]
The lock file is automatically generated by Cargo after a successful build, both for binaries and libraries. As explained in the [FAQ](https://doc.rust-lang.org/cargo/faq.html#why-do-binaries-have-cargolock-in-version-control-but-not-libraries), projects should include Cargo.lock on VCS only for binaries, so that the same dependency versions are used when the package is compiled.

Using Cargo.lock for dependency resolution is necessary because although `cargo-metadata` is able to retrieve all the dependency names, their specific versions are still in semantic versioning format, so it is necessary to check for the latest versions on [crates.io](crates.io) before downloading anything. Another benefit is that it ensures that the user has built the crate at least once before trying to generate an ebuild out of it.

File: README.md [hunk 1]
```diff
@@ -9,7 +9,7 @@
 `cargo ebuild` is a Cargo subcommand that generates an
 [ebuild](https://wiki.gentoo.org/wiki/Ebuild) recipe that uses
 [cargo.eclass](https://gitweb.gentoo.org/repo/gentoo.git/tree/eclass/cargo.eclass)
-to build a Cargo based project for [Gentoo](https://gentoo.org/)
+to build a Cargo based project for [Gentoo](https://gentoo.org).
```
Comment (cardoe) [inline README.md:12]: meh. unrelated change here?

File: README.md [hunk 2]
```diff
@@ -21,133 +21,82 @@ $ cargo install cargo-ebuild
...
-You will first need to get the sources to the crate you want to install.
+Firstly, you need to get the sources to the crate you want to install.
 Your best bet is to search for the crate at [crates.io](https://crates.io)
 and follow the *Repository* link. This should give you the ability to clone
 the repo. Once you have cloned the repo, change into the directory and
 ensure that you checkout the tag corresponding to the version you would like
-to package. Lastly you will execute the `cargo ebuild` command to generate the
-ebuild for that crate.
+to package. The package must contain `Cargo.lock` file to resolve dependency
+versions, otherwise you need to build the project so Cargo can generate it.
```
Comment (cardoe) [inline README.md:30]: So you don't really need to build the package. You can simply do `cargo generate-lockfile`. We can probably even do it within the ebuild command.

Comment (leonardohn) [reply README.md:30]: I think it is a good solution, it can be done when `Cargo.lock` is not found instead of giving an error message.

Comment (leonardohn) [reply README.md:30]: By the way, it doesn't need to check at all, let's just wire `cargo generate-lockfile` as it will check it by itself.

**Comment by leonardohn** [comment]
I have dropped the changes about `Cargo.lock` on README and simplified `lib.rs` considering that `cargo generate-lockfile` is going to be used. Right now the only thing left to do is to invoke it before resoling the dependencies.

**Comment by leonardohn** [comment]
It should be working now, just let me know if you want any changes.

**Comment by leonardohn** [comment]
I've found a problem while testing the changes. `cargo generate-lockfile` is currently overriding `Cargo.lock` if it already exists, so the generated ebuild is failing to compile if any dependency is updated. So I think it would be a better solution to not try to generate it automatically considering that all binaries should come with a Cargo.lock. What do you think @cardoe ?

**Comment by leonardohn** [comment]
Initially I thought that `cargo generate-lockfile` would only generate if the lockfile did not exist, but it actually ignores it and acts like a `cargo update`. I decided to do like you proposed before, if you have any suggestions just let me know.
